File: ios/App/App/TrackApiClient.swift
import Foundation

enum TrackApiClient {
    static func fetchTracking(orderId: String, bearerToken: String?) async throws -> TrackSnapshot {
        let path = "/api/marketplace/orders/\(encode(orderId))/tracking"
        return try await get(path: path, bearerToken: bearerToken)
    }

    static func fetchGuestTracking(orderId: String, phone: String) async throws -> TrackSnapshot {
        let path = "/api/marketplace/orders/\(encode(orderId))/guest-tracking?phone=\(encode(phone))"
        return try await get(path: path, bearerToken: nil)
    }

    private static func get(path: String, bearerToken: String?) async throws -> TrackSnapshot {
        let base = NativeFeatureFlags.apiBaseUrl
        guard let url = URL(string: base + path) else {
            throw URLError(.badURL)
        }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(NativeFeatureFlags.apiVersion, forHTTPHeaderField: "X-Marketplace-API-Version")
        request.setValue(UUID().uuidString, forHTTPHeaderField: "X-Correlation-Id")
        if let bearerToken, !bearerToken.isEmpty {
            request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200...299).contains(code) else {
            throw NSError(domain: "TrackApi", code: code, userInfo: [NSLocalizedDescriptionKey: "tracking_http_\(code)"])
        }
        return try parseEnvelope(data)
    }

    private static func parseEnvelope(_ data: Data) throws -> TrackSnapshot {
        let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        let value: [String: Any]
        if let nested = root["value"] as? [String: Any] {
            value = nested
        } else if root["orderId"] != nil {
            value = root
        } else {
            throw NSError(domain: "TrackApi", code: -1, userInfo: [NSLocalizedDescriptionKey: "tracking_bad_envelope"])
        }
        let timelineRaw = value["timeline"] as? [[String: Any]] ?? []
        let timeline = timelineRaw.map {
            TrackTimelineEvent(
                status: ($0["status"] as? String) ?? "",
                at: ($0["at"] as? String) ?? "",
                message: $0["message"] as? String
            )
        }
        let eta = value["etaMinutes"] as? [String: Any]
        let restaurant = value["restaurant"] as? [String: Any]
        return TrackSnapshot(
            orderId: (value["orderId"] as? String) ?? "",
            orderNumber: (value["orderNumber"] as? String) ?? "",
            status: (value["status"] as? String) ?? "",
            timeline: timeline,
            etaMin: eta?["min"] as? Int,
            etaMax: eta?["max"] as? Int,
            restaurantName: restaurant?["displayName"] as? String
        )
    }

    private static func encode(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value
    }
}

File: android/app/src/main/java/com/bhojanos/customer/track/TrackApiClient.kt
package com.bhojanos.customer.track

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

/**
 * Marketplace tracking client — same headers as hybrid MarketplaceHttpClient.
 */
class TrackApiClient(
    private val apiBaseUrl: String,
    private val apiVersion: String,
) {
    fun fetchTracking(orderId: String, bearerToken: String?): TrackSnapshot {
        val path = "/api/marketplace/orders/${encode(orderId)}/tracking"
        return get(path, bearerToken)
    }

    fun fetchGuestTracking(orderId: String, phone: String): TrackSnapshot {
        val path =
            "/api/marketplace/orders/${encode(orderId)}/guest-tracking?phone=${encode(phone)}"
        return get(path, bearerToken = null)
    }

    private fun get(path: String, bearerToken: String?): TrackSnapshot {
        val url = URL("$apiBaseUrl$path")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "application/json")
            setRequestProperty("X-Marketplace-API-Version", apiVersion)
            setRequestProperty("X-Correlation-Id", UUID.randomUUID().toString())
            if (!bearerToken.isNullOrBlank()) {
                setRequestProperty("Authorization", "Bearer $bearerToken")
            }
        }
        try {
            val code = conn.responseCode
            val body = (if (code in 200..299) conn.inputStream else conn.errorStream)
                ?.bufferedReader()
                ?.use { it.readText() }
                .orEmpty()
            if (code !in 200..299) {
                throw IllegalStateException("tracking_http_$code")
            }
            return parseEnvelope(body)
        } finally {
            conn.disconnect()
        }
    }

    private fun parseEnvelope(body: String): TrackSnapshot {
        val root = JSONObject(body)
        val value = when {
            root.has("value") && !root.isNull("value") -> root.getJSONObject("value")
            root.has("orderId") -> root
            else -> throw IllegalStateException("tracking_bad_envelope")
        }
        val timelineJson = value.optJSONArray("timeline")
        val timeline = buildList {
            if (timelineJson != null) {
                for (i in 0 until timelineJson.length()) {
                    val item = timelineJson.getJSONObject(i)
                    add(
                        TrackTimelineEvent(
                            status = item.optString("status"),
                            at = item.optString("at"),
                            message = item.optString("message").ifBlank { null },
                        ),
                    )
                }
            }
        }
        val eta = value.optJSONObject("etaMinutes")
        val restaurant = value.optJSONObject("restaurant")
        return TrackSnapshot(
            orderId = value.optString("orderId"),
            orderNumber = value.optString("orderNumber"),
            status = value.optString("status"),
            timeline = timeline,
            etaMin = eta?.optInt("min"),
            etaMax = eta?.optInt("max"),
            restaurantName = restaurant?.optString("displayName"),
        )
    }

    private fun encode(value: String): String =
        java.net.URLEncoder.encode(value, Charsets.UTF_8.name())
}
